### ISSUE TITLE
Enable float pixel format on windows for HDR and/or WCG

### ISSUE THREAD

**Issue by miroandel** [glfw/glfw#1907]
This feature request enables to switch from WGL_TYPE_RGBA_ARB to WGL_TYPE_RGBA_FLOAT_ARB (On Windows 10) in order to use higher color bit depths than 10 as well as HDR rendering in non-exclusive mode. This also enables wide color gamut rendering (WCG). Usage for the patch located at the end of this issue:

```c
glfwWindowHint(GLFW_FLOAT_PIXEL_TYPE, GLFW_TRUE);
glfwWindowHint(GLFW_RED_BITS, 16);
glfwWindowHint(GLFW_GREEN_BITS, 16);
glfwWindowHint(GLFW_BLUE_BITS, 16);
glfwWindowHint(GLFW_ALPHA_BITS, 16);
```

Values in the range 0.0f - 1.0f will be in low-dynamic-range while values above 1.0f are in high-dynamic range. The color space is Windows' scRGB where 1.0f corresponds to a brightness of 80 nits, 12.5f corresponds to 1000 nits on a linear scale. More info here: https://docs.microsoft.com/en-us/windows/win32/direct3darticles/high-dynamic-range

For sending color space and HDR metadata to the display, each vendor's API must be used. More info from Nvidia here: 
- https://developer.nvidia.com/high-dynamic-range-display-development
- https://on-demand.gputechconf.com/gtc/2017/presentation/s7394-tom-true-programming-for-high-dynamic-range.pdf

Patch:
```
From bda446768eacf07a4f49dee1b037561e7ebcb369 Mon Sep 17 00:00:00 2001
From: Miro <miroslav@dataton.se>
Date: Fri, 26 Mar 2021 12:15:02 +0100
Subject: [PATCH] Added GLFW_FLOAT_PIXEL_TYPE window creation hint for WGL
 inorder to create float16 pixel-format for HDR and WCG rendering in windows.

---
 include/GLFW/glfw3.h | 5 +++++
 src/context.c        | 6 ++++++
 src/internal.h       | 1 +
 src/wgl_context.c    | 5 ++++-
 src/wgl_context.h    | 1 +
 src/window.c         | 3 +++
 6 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/include/GLFW/glfw3.h b/include/GLFW/glfw3.h
index 7d9dee02..f9caa366 100644
--- a/include/GLFW/glfw3.h
+++ b/include/GLFW/glfw3.h
@@ -982,6 +982,11 @@ extern "C" {
  *  Framebuffer double buffering [hint](@ref GLFW_DOUBLEBUFFER).
  */
 #define GLFW_DOUBLEBUFFER           0x00021010
+ /*! @brief Framebuffer pixel format float type hint.
+  *
+  *  Framebuffer pixel format float type [hint](@ref GLFW_FLOAT_PIXEL_TYPE).
+  */
+#define GLFW_FLOAT_PIXEL_TYPE       0x00021011
 
 /*! @brief Context client API hint and attribute.
  *
diff --git a/src/context.c b/src/context.c
index 48311e5f..b8bf5dce 100644
--- a/src/context.c
+++ b/src/context.c
@@ -202,6 +202,12 @@ const _GLFWfbconfig* _glfwChooseFBConfig(const _GLFWfbconfig* desired,
             continue;
         }
 
+        if (desired->floatPixelType != current->floatPixelType)
+        {
+            // Float pixel type is a hard constraint
+            continue;
+        }
+
         // Count number of missing buffers
         {
             missing = 0;
diff --git a/src/internal.h b/src/internal.h
index f87c2f93..083228f4 100644
--- a/src/internal.h
+++ b/src/internal.h
@@ -335,6 +335,7 @@ struct _GLFWfbconfig
     GLFWbool    sRGB;
     GLFWbool    doublebuffer;
     GLFWbool    transparent;
+    GLFWbool    floatPixelType;
     uintptr_t   handle;
 };
 
diff --git a/src/wgl_context.c b/src/wgl_context.c
index 4f9a6ffe..952a9343 100644
--- a/src/wgl_context.c
+++ b/src/wgl_context.c
@@ -159,7 +159,8 @@ static int choosePixelFormat(_GLFWwindow* window,
                 continue;
             }
 
-            if (findAttribValue(WGL_PIXEL_TYPE_ARB) != WGL_TYPE_RGBA_ARB)
+            if (findAttribValue(WGL_PIXEL_TYPE_ARB) != WGL_TYPE_RGBA_ARB &&
+                findAttribValue(WGL_PIXEL_TYPE_ARB) != WGL_TYPE_RGBA_FLOAT_ARB)
                 continue;
 
             if (findAttribValue(WGL_ACCELERATION_ARB) == WGL_NO_ACCELERATION_ARB)
@@ -184,6 +185,8 @@ static int choosePixelFormat(_GLFWwindow* window,
                 u->stereo = GLFW_TRUE;
             if (findAttribValue(WGL_DOUBLE_BUFFER_ARB))
                 u->doublebuffer = GLFW_TRUE;
+            if (findAttribValue(WGL_PIXEL_TYPE_ARB) == WGL_TYPE_RGBA_FLOAT_ARB)
+                u->floatPixelType = GLFW_TRUE;
 
             if (_glfw.wgl.ARB_multisample)
                 u->samples = findAttribValue(WGL_SAMPLES_ARB);
diff --git a/src/wgl_context.h b/src/wgl_context.h
index 2cf7e4e5..05e70cd8 100644
--- a/src/wgl_context.h
+++ b/src/wgl_context.h
@@ -30,6 +30,7 @@
 #define WGL_DRAW_TO_WINDOW_ARB 0x2001
 #define WGL_PIXEL_TYPE_ARB 0x2013
 #define WGL_TYPE_RGBA_ARB 0x202b
+#define WGL_TYPE_RGBA_FLOAT_ARB 0x21A0
 #define WGL_ACCELERATION_ARB 0x2003
 #define WGL_NO_ACCELERATION_ARB 0x2025
 #define WGL_RED_BITS_ARB 0x2015
diff --git a/src/window.c b/src/window.c
index d2196e2f..982472a9 100644
--- a/src/window.c
+++ b/src/window.c
@@ -343,6 +343,9 @@ GLFWAPI void glfwWindowHint(int hint, int value)
         case GLFW_SRGB_CAPABLE:
             _glfw.hints.framebuffer.sRGB = value ? GLFW_TRUE : GLFW_FALSE;
             return;
+        case GLFW_FLOAT_PIXEL_TYPE:
+            _glfw.hints.framebuffer.floatPixelType = value ? GLFW_TRUE : GLFW_FALSE;
+            return;
         case GLFW_RESIZABLE:
             _glfw.hints.window.resizable = value ? GLFW_TRUE : GLFW_FALSE;
             return;
-- 
2.31.1.windows.1

```

**Comment by eddiejames** [comment]
Thanks for this! Is there any reason this hasn't been opened as a PR and included in GLFW?

**Comment by miroandel** [comment]
I can have a look on how to make a PR out of it. It probably needs rebasing to master. Also half-float buffer works only on Windows at the moment as I understand it. There might need to be some error handling when trying to use GLFW_FLOAT_PIXEL_TYPE on other OS:es/windowing systems.

I'm also a but unsure about if there are any rules when creating new defines. I simply set a value that wasn't used.

However, this code has been used successfully on both Nvidia GPUs, Intel GPUs, and hybrid versions between those. It probably works on AMD too but hasn't been tested. Windows is handling the static HDR meta data but it's possible to override with Nvidias API for example.

**Comment by eddiejames** [comment]
OK, yea I was just testing it yesterday. I'm on AMD. It worked, but strangely, I could only get a BGRA buffer. I thought to add some code to check the bit shifts, but they were actually all BGRA.

**Comment by meshula** [comment]
I'd like to get fp16 and other extended formats on macOS as well ~ grepping around in the code I think float pixel formats haven't been implemented since this issue was posted? Just want to double check that I haven't overlooked something.

**Comment by Ext3h** [comment]
> Values in the range 0.0f - 1.0f will be in low-dynamic-range while values above 1.0f are in high-dynamic range

That's technically incorrect. When using an sRGB render target, 1.0f is the user defined preferred white point when targeting an HDR display (that behaves like LDR content, and that 1.0f is usually far off from 80 nits). But when using scRGB render target, 1.0f is reduced to 80 nits, and you have to query the native system APIs to get the preferred brightness of "reference white" as a multiplier just in order to get LDR content to show as expected again.

So you can't just switch the default behavior and expect applications to "just work".

> For sending color space and HDR metadata to the display

Forget about that part for now. That's the least of your concerns. Not to mention that most displays ignore this information anyway, as it can be trivially reconstructed from the image content. Plus you are clashing with Windows here, which has already opted to take control, and has (for very good reasons) stated towards to the display that it will produce content which ***exactly*** fits the native capabilities of the display.

More important is the part where you query the effective color gamut and brightness range. Everything in that range is safe to use, and will behave somewhat as expected. This is what the final applications tone mapping needs to yield.

Everything outside that range is subject to "smart optimizations" the display vendor may have applied. Do not go this route, you will encounter highly unexpected effects such as the display randomly adjusting the brightness, deviating from the previously absolute scale, just in order to still fit the highlights into it's own range. If you do this, you will face displays (thanks for nothing, BenQ) trying to do the tone mapping in your stead, and the result will suck. You won't notice this specific issue if your reference system is a 10bit wide gamut display with a color profile, in that case Windows had already taken care of properly clipping.

TLDR: Make sure the content is clipped to the displays capabilities, or UB occurs.

**Comment by audaki** [comment]
We got some new OpenGL HDR implementations (GLFW fork for KDE6 HDR, and Oolite who directly output OpenGL HDR in windows without DXGI, also using a float pixel format), referenced from here: https://github.com/glfw/glfw/issues/890#issuecomment-2154960317

**Comment by zezba9000** [comment]
Is the problem OpenGL doesn't support HDR SwapChain on Windows and doesn't support HDR content on Linux?
Or just that GLFW is being used by Blender and that doesn't support HDR correctly on these Win/Lin?

**Comment by MiroslavAndel** [comment]
@zezba9000, In Windows OpenGL doesn't really know about HDR. In WGL instead of creating an integer pixelformat (WGL_TYPE_RGBA_ARB) a floating point pixel format can be created instead (WGL_TYPE_RGBA_FLOAT_ARB). This doesn't clamp the values from the fragment shader to [0.0, 1.0]. If HDR is enabled on that monitor in windows then the pixel data is interpreted as scRGB with a range of [0.0, 125.0] by the desktop manager when compositing the buffer to the screen. Then it's up to the desktop manager to tonemap, clip or whatever based on the HDR metadata from the monitor.
To get the values in nits (or cd/m^2) you can scale the values by 80 resulting in that a value of 1.0 in the buffer maps to 80 nits, while 12.5 maps to 1000 nits and the max value maps to 10000 nits which is the limit for Rec. 2100 using the PQ OETF.

To get HDR in a GLWF app if GLFW supports a floating point buffer is to be able to render values above 1.0 using a linear mapping instead of sRGB or gamma 2.2.

**Comment by zezba9000** [comment]
If you can just use WGL format, then whats this bit about using DXGI to create the texture? Are the HDR ranges different / correct using DXGI to create the texture buffer?

**Comment by Ext3h** [comment]
> If you can just use WGL format, then whats this bit about using DXGI to create the texture? Are the HDR ranges different / correct using DXGI to create the texture buffer?

It's doing the same thing. At least it's supposed to, except that it's the driver doing it for you behind the scenes with WGL. And there is/was the catch that the Nvidia driver is doing something wrong, which results in stability issues when trying to do it via WGL. Which is why the explicit invocation of the interop on the application side is preferred.

In both cases, you absolutely must use the SDR white level query to correctly map SDR / sRGB content to the scRGB buffer. That is never plug&play.

**Comment by mirh** [comment]
Yeah, *presumably* it's the drivers doing the magic under the hood, not the desktop manager . 
According to https://github.com/IrisShaders/Iris/issues/2282 this preceded the official introduction of DXGI presentation (in fact, that was probably just exposing an internal knob they had already built exactly for this HDR scenario).
But it's a bit of a mystery just how far back in time that's supported (or, say, if Intel gpus would work without the explicit interop).

**Comment by goofyseeker311** [comment]
Any progress on this?

**Comment by corporateshark** [comment]
Any updates?

**Comment by mirh** [comment]
FWIW https://github.com/rrtt217/glfw/pull/1 added support for the native DXGI swapchain
